### PR TITLE
dev: bump cairo compiler

### PR DESCRIPTION
**Pull request by greged93**
# Bump cairo compiler 

## Description
Bumps the cairo compiler dependency to 2.4.1.

Description of the pull request changes and motivation.

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

**Comment by greged93** [comment]
Copied the benchmark results since they can't get posted on the PR because it originates from another repo (i.e. my fork):

# Benchmarking results
## Benchmark for program `factorial_2M`

<details><summary>Open benchmarks</summary>

<br>

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `Cairo-vm (Rust, Cairo 1)` | 17.937 ± 0.076 | 17.810 | 18.106 | 37.00 ± 0.16 |
| `cairo-native (JIT MLIR ORC Engine)` | 1.403 ± 0.004 | 1.395 | 1.408 | 2.89 ± 0.01 |
| `cairo-native (AOT Native binary)` | 0.700 ± 0.002 | 0.697 | 0.703 | 1.44 ± 0.00 |
| `cairo-native (AOT Native binary with host CPU features, march=native)` | 0.485 ± 0.000 | 0.484 | 0.485 | 1.00 |

</details>

## Benchmark for program `fib_2M`

<details><summary>Open benchmarks</summary>

<br>

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `Cairo-vm (Rust, Cairo 1)` | 17.473 ± 0.044 | 17.411 | 17.543 | 598.27 ± 2.43 |
| `cairo-native (JIT MLIR ORC Engine)` | 0.948 ± 0.010 | 0.932 | 0.965 | 32.46 ± 0.36 |
| `cairo-native (AOT Native binary)` | 0.041 ± 0.000 | 0.041 | 0.041 | 1.41 ± 0.01 |
| `cairo-native (AOT Native binary with host CPU features, march=native)` | 0.029 ± 0.000 | 0.029 | 0.029 | 1.00 |

</details>

## Benchmark for program `logistic_map`

<details><summary>Open benchmarks</summary>

<br>

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `Cairo-vm (Rust, Cairo 1)` | 1.651 ± 0.013 | 1.635 | 1.676 | 23.64 ± 0.21 |
| `cairo-native (JIT MLIR ORC Engine)` | 0.998 ± 0.007 | 0.988 | 1.006 | 14.29 ± 0.12 |
| `cairo-native (AOT Native binary)` | 0.109 ± 0.000 | 0.109 | 0.110 | 1.57 ± 0.01 |
| `cairo-native (AOT Native binary with host CPU features, march=native)` | 0.070 ± 0.000 | 0.070 | 0.071 | 1.00 |

</details>

File: Makefile [hunk 1]
```diff
@@ -5,7 +5,7 @@
 #
 
 UNAME := $(shell uname)
-CAIRO_2_VERSION=2.3.1
+CAIRO_2_VERSION=2.4.1
```
Comment (edg-l) [inline Makefile:8]: can you change this to 2.4.2?

File: Cargo.toml [hunk 2]
```diff
@@ -27,13 +27,13 @@ with-runtime = ["cairo-native-runtime"]
...
 [dependencies]
 bumpalo = "3.14"
-cairo-lang-compiler = "2.3.1"
-cairo-lang-defs = "2.3.1"
-cairo-lang-diagnostics = "2.3.1"
-cairo-lang-filesystem = "2.3.1"
-cairo-lang-lowering = "2.3.1"
-cairo-lang-sierra = "2.3.1"
-cairo-lang-sierra-generator = "2.3.1"
+cairo-lang-compiler = "2.4.0"
```
Comment (edg-l) [inline Cargo.toml:30]: is there a reason you are using 2.4.0 instead of 2.4.2?

Comment (greged93) [reply Cargo.toml:30]: overlooked this dep, will update. The reason I was on `2.4.1` was because the contract I would like to test cairo-native against was compiled on 2.4.1, so didn't realize there was a `2.4.2` already. Will update!

**Review by igaray — APPROVED** [review]

**Comment by greged93** [comment]
@edg-l I've been trying to implement the `match_nullable_snapshot` as well and I'm facing a weird behaviour, maybe you can see what I'm doing wrong (if not I will just push the compiler version updates):
- in the build for `snapshot_take`, I print the pointer and the underlying value. I use the below code:
```rust
let res = entry
    .append_operation(llvm::load(
        context,
        entry.argument(0)?.into(),
        IntegerType::new(context, 8).into(),
        location,
        LoadStoreOptions::new().align(Some(IntegerAttribute::new(
            8,
            IntegerType::new(context, 64).into(),
        ))),
    ))
    .result(0)?
    .into();
metadata.get_mut::<DebugUtils>().unwrap().print_pointer(
    context,
    helper,
    entry,
    entry.argument(0)?.into(),
    location,
)?;
metadata
    .get_mut::<DebugUtils>()
    .unwrap()
    .print_i8(context, helper, entry, res, location)?;
```
This gives me the below when taking the snap:
```
[DEBUG] 0x000000010fca3440
[DEBUG] 2
```
- In the build for `match_nullable_snapshot` which I added, I use the following to print the underlying pointer value:
```rust
let (inner_ty, inner_layout) =
    registry.build_type_with_layout(context, helper, registry, metadata, &info.ty)?;
let arg_u8 = entry
    .append_operation(llvm::load(
        context,
        entry.argument(0)?.into(),
        inner_ty,
        location,
        LoadStoreOptions::new().align(Some(IntegerAttribute::new(
            inner_layout.align() as i64,
            IntegerType::new(context, 64).into(),
        ))),
    ))
    .result(0)?
    .into();
metadata
    .get_mut::<DebugUtils>()
    .unwrap()
    .print_pointer(context, helper, entry, entry.argument(0)?.into(), location)?;
metadata
    .get_mut::<DebugUtils>()
    .unwrap()
    .print_i8(context, helper, entry, arg_u8, location)?;
```
In this case, I get
```
[DEBUG] 0x0000000142e17180
[DEBUG] 0
```
Let me know if this is enough info for you, if not I'll just bump the compiler updates and can push the code in another PR.

**Comment by edg-l** [comment]
I havent looked at match_nullable_snapshot but if the argument is a nullable, then its a pointer, which maybe is a pointer to a `info.ty` itself and not the inner type of the enum?

**Comment by azteca1998** [comment]
Hello, and thanks for your PR.

I think your issue with `match_nullable_snapshot` may be caused by a refactor we did last month to improve the interface between C and Cairo when enums are involved. We still have to update the docs, but basically if the enum meets some criterion, then the enum is passed by reference (a stack-allocated pointer) instead of by value.

Since you're not showing the enum you're testing I'll assume it is indeed one that requires stack allocation, therefore you'd get a `!llvm.ptr<!llvm.ptr>` instead of a `!llvm.ptr<the actual enum type>` as an argument to the `match_nullable_snapshot` libfunc builder.

The fix is to:
  1. Check if the enum is stack allocated.
  2. If it is, then load a pointer.
  3. Load the enum data from the previous pointer if loaded, or the argument otherwise.

Please let us know if this fixes the weird behavior you found.

**Comment by greged93** [comment]
Hey @azteca1998, I had some time to check the code and I think the error comes from desnapping a Nullable. Not sure why but the `*` operator seems to just replace the value in the Nullable by 0. Below is a test I made that fails to pass:
```rust
#[test]
fn run_desnap_nullable() {
    let program = load_cairo!(
        use nullable::NullableTrait;
        use nullable::FromNullableResult;
        use nullable::match_nullable;

        fn run_test(x: u8) -> u8 {
            let a = @NullableTrait::new(x);
            let b = *a;
            let c = match match_nullable(b) {
                FromNullableResult::Null(_) => 99_u8,
                FromNullableResult::NotNull(value) => value.unbox()
            };
            c
        }
    );

    run_program_assert_output(&program, "run_test", &[4u8.into()], 4u8.into());
}
```
Fails with
```
thread 'libfuncs::nullable::test::run_desnap_nullable' panicked at 'assertion failed: `(left == right)`

Diff < left / right > :
 Uint8(
<    0,
>    4,
 )
```

This test compiles and passes using scarb.

Also compiled this very simple cairo code to check what a desnap operator gets compiled to in sierra:
```rust
fn test() {
    let c = @NullableTrait::new(1_u8);
    let d = *c;
    match match_nullable(d) {
        FromNullableResult::Null => 1_u8,
        FromNullableResult::NotNull(_) => 2_u8
    };
}
```
Correct me if I am wrong but the desnap compiles to a `drop` libfunc and then a `rename` libfunc. The issue with desnapping is that currently the `drop` implementation frees the pointer by calling the "free" function of the `ReallocBindingsMeta` (not sure what the function does, couldn't find info) which in turn I think causes me to have 0 as the underlying value of the pointer.
https://github.com/lambdaclass/cairo_native/blob/16b36c8fb609c5a16ca646b9784fac291994f80f/src/types.rs#L744

Let me know how you want to proceed, I think the `drop` libfunc needs to be updated for a `Nullable` as far as I understood.

**Comment by azteca1998** [comment]
I think I see the problem now. It has nothing to do with the enums.

Basically, a snapshot in the VM is just a pointer to a segment. This works because memory in the VM cannot be reassigned. However in `cairo-native` we don't have segments, therefore we need to copy the entire value. This is done in `snapshot_take`, where there's a section of the code in charge of that (the one that uses `SnapshotClonesMeta`). The problem is that there's no entry in `SnapshotClonesMeta` for nullables yet, only for arrays.

You could try to add something like [this](https://github.com/lambdaclass/cairo_native/blob/main/src/types/array.rs#L62-L71) but in the [nullable type builder](https://github.com/lambdaclass/cairo_native/blob/main/src/types/nullable.rs#L26-L40). You'd need to implement the clone mechanism ([example here](https://github.com/lambdaclass/cairo_native/blob/main/src/types/nullable.rs#L26-L40)) too, which should be just a `realloc` followed by an inline `memcpy`.

But I think that's enough work to justify a separate PR. Please let me know if you want to try it yourself, meanwhile I'll just open an issue.

**Comment by greged93** [comment]
Thanks for the detailed explanation! Will push compiler updates here. I don't think I will have time to work on the `Nullable` issue unfortunately :/ Thanks for your time!

**Comment by azteca1998** [comment]
Understood, I'll try to have a fix for your issue by today so that your tests pass.

Also, I think you may have forgotten to change the cairo version to `2.4.2` in the `Makefile`.

File: src/libfuncs/nullable.rs [hunk 3]
```diff
@@ -173,6 +173,26 @@ where
...
+    metadata: &mut MetadataStorage,
+    info: &SignatureAndTypeConcreteLibfunc,
+) -> Result<()>
+where
+    TType: GenericType,
+    TLibfunc: GenericLibfunc,
+    <TType as GenericType>::Concrete: TypeBuilder<TType, TLibfunc, Error = CoreTypeBuilderError>,
+    <TLibfunc as GenericLibfunc>::Concrete: LibfuncBuilder<TType, TLibfunc, Error = Error>,
+{
+    build_match_nullable(context, registry, entry, location, helper, metadata, info)
```
Comment (greged93) [inline src/libfuncs/nullable.rs:193]: I'm not 100% sure about this, but since the input of the libfunc is a `@Nullable<T>`, does it make since to say it's equivalent to a `Nullable<T>` since `snapshot_take` just copies the pointer to the value in the MLIR implementation? Same can be said about `Box<T>` and `Box<@T>`.

**Review by edg-l — APPROVED** [review]

File: src/libfuncs/nullable.rs (hunk 3 above)
Comment (azteca1998) [reply src/libfuncs/nullable.rs:193]: Yes, `snapshot_take` is (or should be) equivalent to a `.clone()` in rust, not a plain copy since otherwise we'd get double free and other undesireable side effects. Because of this, any type containing snapshots should be internally equivalent to the same type but with snapshots removed.

TLDR: It should be correct.

**Review by azteca1998 — APPROVED** [review]

**Review by igaray — APPROVED** [review]
